### PR TITLE
feat: allow client options to be set in magics context

### LINKED ISSUE #160: The magics API do not support setting client options
#### Environment details

  - OS type and version: jupyter notebook
  - Python version: Python 3.7.3
  - pip version: pip 20.1.1
  - google-cloud-bigquery version: 1.25.0

#### Steps to reproduce
This is a missing option that do exists in google bigquery client - "https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client"
But is missing from "https://github.com/googleapis/python-bigquery/blob/master/google/cloud/bigquery/magics.py"

#### Code example

```python
client = bigquery.Client(
        project=project,
        credentials=context.credentials,
        default_query_job_config=context.default_query_job_config,
        client_info=client_info.ClientInfo(user_agent=IPYTHON_USER_AGENT),
    )
```
Note the the client_option is missing in creating the client in the code above

The patch to apply to fix this flow:
```diff
diff --git a/google/cloud/bigquery/magics.py b/google/cloud/bigquery/magics.py
index 7128e32..00e26d0 100644
--- a/google/cloud/bigquery/magics.py
+++ b/google/cloud/bigquery/magics.py
@@ -140,6 +140,10 @@
         Out[7]:     num
            ...: -------
            ...: 0    17
+        Executing query with job ID: bf633912-af2c-4780-b568-5d868058632b
+        Query executing: 2.61s
+        Query complete after 2.92s
+
 """
 
 from __future__ import print_function
@@ -183,6 +187,7 @@ class Context(object):
         self._project = None
         self._connection = None
         self._default_query_job_config = bigquery.QueryJobConfig()
+        self._client_options = google.api_core.client_options.ClientOptions()
 
     @property
     def credentials(self):
@@ -244,6 +249,30 @@ class Context(object):
     def project(self, value):
         self._project = value
 
+    @property
+    def client_options(self):
+        """ google.api_core.client_options.ClientOptions: client options to be used through IPython
+        Magics
+
+        Note::
+            The client options does not need to be explicitly defined if you have an
+            no special network conenctions are required. Normally you would
+            be using https://www.googleapis.com/ end point
+
+        Example:
+            Manually setting the endpoint:
+
+            >>> from google.cloud.bigquery import magics
+            >>> client_option = {}
+            >>> client_option['api_endpoint'] = "https://some.special.url"
+            >>> magics.context.client_options = client_option
+        """
+        return self._client_options
+
+    @client_options.setter
+    def client_options(self, value):
+        self._client_options = value
+
     @property
     def default_query_job_config(self):
         """google.cloud.bigquery.job.QueryJobConfig: Default job
@@ -459,6 +488,18 @@ def _create_dataset_if_necessary(client, dataset_id):
         "name (ex. $my_dict_var)."
     ),
 )
+@magic_arguments.argument(
+    "--options",
+    nargs="+",
+    default=None,
+    help=(
+        "Parameters to format the query string. If present, the --options "
+        "flag should be followed by a string representation of a dictionary "
+        "in the format (Union[google.api_core.client_options.ClientOptions, Dict])"
+        "or a reference to a dictionary in the same format. The dictionary "
+        "key/value matches does at google.api_core.client_options.ClientOptions"
+    ),
+)
 def _cell_magic(line, query):
     """Underlying function for bigquery cell magic
 
@@ -501,6 +542,7 @@ def _cell_magic(line, query):
         credentials=context.credentials,
         default_query_job_config=context.default_query_job_config,
         client_info=client_info.ClientInfo(user_agent=IPYTHON_USER_AGENT),
+        client_options=context.client_options
     )
     if context._connection:
         client._connection = context._connection
```

### PR DESCRIPTION
**Pull request by cguardia**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #160 🦕

### CODE REVIEW

File: google/cloud/bigquery/magics/magics.py [hunk 1]
```diff
@@ -410,6 +437,15 @@ def _create_dataset_if_necessary(client, dataset_id):
         "Standard SQL if this argument is not used."
     ),
 )
+@magic_arguments.argument(
+    "--api_endpoint",
+    type=str,
+    default=None,
+    help=(
+        "The desired API endpoint, e.g., compute.googlepis.com. Defaults to this "
```
Comment (tswast) [inline google/cloud/bigquery/magics/magics.py:445]: Let's make the example bigquery-specific

```suggestion
        "The desired API endpoint, e.g., bigquery.googlepis.com. Defaults to this "
```

File: google/cloud/bigquery/magics/magics.py [hunk 2]
```diff
@@ -410,6 +437,15 @@ def _create_dataset_if_necessary(client, dataset_id):
         "Standard SQL if this argument is not used."
     ),
 )
+@magic_arguments.argument(
+    "--api_endpoint",
```
Comment (tswast) [inline google/cloud/bigquery/magics/magics.py:441]: We might actually want two API endpoints, one for the BigQuery client and one for the BigQuery Storage client.

File: google/cloud/bigquery/magics/magics.py [hunk 3]
```diff
@@ -239,6 +242,30 @@ def project(self):
     def project(self, value):
         self._project = value
 
+    @property
+    def client_options(self):
```
Comment (tswast) [inline google/cloud/bigquery/magics/magics.py:246]: We might actually want two client options, one for the BigQuery client and one for the BigQuery Storage client.

**Comment by cguardia** [comment]
@tswast OK, added both client options now.

**Review by tswast — CHANGES_REQUESTED** [review]
Looking good! Just one nit about naming

File: google/cloud/bigquery/magics/magics.py [hunk 4]
```diff
@@ -243,28 +244,52 @@ def project(self, value):
...
 
-    @client_options.setter
-    def client_options(self, value):
-        self._client_options = value
+    @bigquery_client_options.setter
+    def bigquery_client_options(self, value):
+        self._bigquery_client_options = value
+
+    @property
+    def storage_client_options(self):
```
[the remark this answers is not in the document]
Comment (cguardia) [reply google/cloud/bigquery/magics/magics.py]: Right. Done.

**Review by tswast — APPROVED** [review]

**Comment by boaz-work** [comment]
Thanks

**Pull request merged**